### ISSUE TITLE
LSTM Layer crashes the workstation with num_layers >10

### ISSUE THREAD

**Issue by chahuja** [pytorch/pytorch#877]
I implemented a basic character level langauge model using LSTMs, but the workstation crashes if the number of layers are high (~10 or more).

**Comment by soumith** [comment]
if you give a script to reproduce this, happy to take a look. otherwise, it's hard to know how to identify and fix the issue.

**Comment by chahuja** [comment]
Following is the script I use. You would need a text file(any text file with a lot of characters will do) of the name `input.txt` in a directory named `dataset`. Check the variable `data_file `.

Currently the variable `num_layers` is set to 10. This causes my system to crash. 

I use the conda version of pytorch and  TitanX (12GB) for running this code.
```
import numpy as np
import torch
from torch.autograd import Variable
import pdb
from tqdm import tqdm

# Data Loader
class BatchGenerator(object):
  def __init__(self, text, batch_size, num_unrolling, vocabulary_size, char2id):
    self._text = text
    self._batch_size = batch_size
    self._num_unrolling = num_unrolling
    self._vocabulary_size = vocabulary_size
    self._char2id = char2id

  def __call__(self):
    ## cut extra data at the end
    text_len = (len(self._text)//self._batch_size) * self._batch_size
    text_len = (((text_len/self._batch_size -1) // self._num_unrolling) * self._num_unrolling + 1) * self._batch_size
    self._text = self._text[:text_len]

    ## create batches
    batches = list()
    for step in range(self._num_unrolling + 1):
      segment = text_len/self._batch_size
      arr_len = text_len/self._num_unrolling
      arr = np.zeros((arr_len, self._vocabulary_size))
      sample_num = 0
      for c in range(step,segment-1, self._num_unrolling):
        for b in range(self._batch_size):
          arr[sample_num, self._char2id(self._text[b*segment+c])] = 1
          sample_num+=1
      batches.append(arr)

    return batches

class Data(object):
  def __init__(self, filename, valid_frac=0.1, batch_size=100, num_unrolling=10, criterion='char', delimiter=None):
    '''
    Data handler for character/word based models.
    criterion: char/word based spliting
    '''

    self.unk_count = 0

    if(delimiter==None):
      if(criterion=='word'):
        ## search for existing tokenised file
        if(not os.path.isfile(filename + '.tokens')):
          subprocess.call('./tokenizer.perl -l en -no-escape' + ' < ' + filename + ' > ' + filename+'.tokens', shell=True)

        filename += '.tokens'
        delimiter = ' '
      elif(criterion=='char'):
        delimiter = ''
      else:
        assert (0), 'Incorrect criterion. Choose from char or word'
    
    ## read text from a given file
    text = self.read_data(filename)
    if(criterion=='word'):
      ## tokenzer.perl does not put spaces between escape sequences
      text = text.replace('\n',' \n ').replace('\n  \n','\n \n') ## hardcoded for shakespere dataset

    if(delimiter == ''):
      text = list(text)
    else:
      text=text.split(delimiter)

    self.delimiter = delimiter
      
    ## split text into train and test
    train_text, valid_text, test_text = self.split_data(text, valid_frac)

    ## create vocabulary based on the data
    self.vocab, self.rev_vocab, self.vocabulary_size = self.create_vocab(train_text)

    train = BatchGenerator(train_text, batch_size, num_unrolling, self.vocabulary_size, self.char2id)
    val = BatchGenerator(valid_text, batch_size, 1, self.vocabulary_size, self.char2id)
    test = BatchGenerator(test_text, batch_size, 1, self.vocabulary_size, self.char2id)
    self.train_batches = train()
    self.val_batches = val()
    print('Unknown Tokens in val-set: %d') %(self.unk_count)
    self.unk_count = 0
    self.test_batches = test()
    print('Unknown Tokens in test-set: %d') %(self.unk_count)
    
  ## Read file as a single string
  def read_data(self, filename):
    f = open(filename, 'rb')
    data=f.read()
    f.close()
    return data

  ## Split given string into train and test
  def split_data(self, text, valid_frac):
    """Split the given data into train and validation based on
    the valid_frac"""
    valid_size = int(np.ceil(len(text)*valid_frac))
    test_size = int(np.ceil(len(text)*valid_frac*3)) ## hardcoded to 3 times valid frac
    valid_text = text[-valid_size:]
    test_text = text[-valid_size-test_size:-valid_size]
    train_text = text[:-valid_size-test_size]
    train_size = len(train_text)
    print(train_size, train_text[:64])
    print(valid_size, valid_text[:64])
    print(test_size, test_text[:64])
    return train_text, valid_text, test_text

  ## Given a string find a character/word based vocabulary
  def create_vocab(self, text):
    """Create a vocabulary for a given text and return 
    vocab, rev_vocab and vocabulary size
    """
    
    vocabulary_size = 1
    vocab = {'_UNK':0}
    rev_vocab = ['_UNK']
    for char in text:
      if char not in vocab.keys():
        vocab[char] = vocabulary_size
        rev_vocab += [char]
        vocabulary_size += 1
    print('Vocabulary Size: %d') % (vocabulary_size)
    return vocab, rev_vocab, vocabulary_size

  def char2id(self,char):
    """Convert a character to its correspoding id in the dictionary"""
    if char in self.vocab.keys():
      return self.vocab[char]
    else:
      self.unk_count += 1
      return 0

  def id2char(self,dictid):
    """Convert id(int) to the corresponding character"""
    if dictid >=0 :
      return self.rev_vocab[dictid]
    else:
      print ('Invalid ID: %d' % dictid)
      return 0

  def characters(self,probabilities):
    """Turn a 1-hot encoding or a probability distribution over the possible
    characters back into its (most likely) character representation."""


## Variables
data_file = './dataset/input.txt'
num_unrolling = 100
batch_size = 100
num_layers = 10
rnn_size = 418
NUM_EPOCHS = 100

data = Data(data_file, num_unrolling=num_unrolling, batch_size=batch_size)

train = torch.Tensor(np.stack(np.asarray(data.train_batches, dtype=np.int), axis=1)[:,:-1,:])
y_t = torch.LongTensor(np.argmax(np.stack(np.asarray(data.train_batches, dtype=np.int), axis=1)[:,1:,:], axis=2))

class seq2seq(torch.nn.Module):
  def __init__(self, vocab_size, batch_size, hidden_state, num_layers):
    super(seq2seq,self).__init__()
    self.vocab_size = vocab_size
    self.batch_size = batch_size
    self.hidden_state = hidden_state
    self.num_layers = num_layers
    self.model = torch.nn.LSTM(input_size=vocab_size, hidden_size=hidden_state, num_layers=num_layers, batch_first=True)
    self.predict = torch.nn.Linear(hidden_state,vocab_size)
    #self.Wo = Variable(torch.FloatTensor(np.random.uniform(0.01,-0.01,size=(self.vocab_size, self.hidden_state))),requires_grad=True).cuda()
    #self.bo = Variable(torch.FloatTensor(np.zeros((self.vocab_size))),requires_grad=True).cuda()
    
  def forward(self,x):
    zeros = torch.autograd.Variable(torch.zeros(*(self.num_layers, self.batch_size, self.hidden_state)))
    h, _ = self.model(x, None)
    #pdb.set_trace()
    #y = h.contiguous().view(-1,self.hidden_state).mm(self.Wo) + self.bo
    #y = torch.nn.functional.linear(h.contiguous().view(-1,self.hidden_state),self.Wo,self.bo)
    y = self.predict(h.contiguous().view(-1,self.hidden_state))
    return y
    

model = seq2seq(vocab_size=train.size(2), batch_size=batch_size, hidden_state=rnn_size, num_layers=num_layers)
model = model.cuda()


criterion = torch.nn.CrossEntropyLoss() # use a Classification Cross-Entropy loss
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)


indices = range(0,train.size(0)+1,batch_size)
for i in tqdm(range(NUM_EPOCHS)):
  running_loss = 0.0
  for i, (start,end) in enumerate(zip(indices[:-1],indices[1:])):
    inputs, labels = torch.autograd.Variable(train[start:end,:,:]).cuda(), torch.autograd.Variable(y_t[start:end,:]).cuda()
    
    # zero the parameter gradients
    optimizer.zero_grad()
    outputs = model(inputs)
    
    loss = criterion(outputs, labels.view(-1))
    
    loss.backward()        
    running_loss += loss.data[0]
    optimizer.step()
  tqdm.write('Loss:%f'%(running_loss/(i+1)))

```

**Comment by apaszke** [comment]
It would be helpful if you could simplify your example and paste it into a gist. Debugging scripts is pretty time consuming, so the less code you leave the better.

**Comment by chahuja** [comment]
[Gist](https://gist.github.com/chahuja/e31f1ba934598126c9dd39d2e167c770)

**Comment by apaszke** [comment]
That's still quite long. We probably don't need data generation and all that. Just replace it with `torch.randn` and similar. It'd be the best if you could manage to compress the script into 30-40 lines.

**Comment by chahuja** [comment]
Reduced the size to its minimal form. 
[Gist](https://gist.github.com/chahuja/e31f1ba934598126c9dd39d2e167c770)

**Comment by apaszke** [comment]
Is it the memory that's blowing up? It seems to be related to #914.

**Comment by chahuja** [comment]
Yes, I think excessive memory usage covers this up as stated in #914. Although, it might be good to have exit statements for excess memory usage in general, preventing the crash.

**Comment by apaszke** [comment]
I'm closing this, should be fixed once #914 is resolved.
